llms/gguf_llm/models.py

```diff
@@ -0,0 +1,363 @@
+    def __init__(self, metadata):
+        def parse_token(token):
+            if len(token) == 6 and token.startswith("<0x") and token.endswith(">"):
+                return chr(int(token[3:5], 16))
+            return token
+
+        # TODO: do we need scores and token type?
+        # sores = metadata["tokenizer.ggml.scores"]
+        # token_types = metadata["tokenizer.ggml.token_type"]
+        vocab = {
+            parse_token(t): i for i, t in enumerate(metadata["tokenizer.ggml.tokens"])
+        }
+        merges = [
+            tuple(m.split(" ")) for m in metadata.get("tokenizer.ggml.merges", [])
+        ]
+        model = BPE(vocab, merges, byte_fallback=True)
+        self._tokenizer = Tokenizer(model)
+        self._bos_token_id = metadata["tokenizer.ggml.bos_token_id"].item()
+        self._eos_token_id = metadata["tokenizer.ggml.eos_token_id"].item()
+
+    def encode(self, s: str) -> mx.array:
+        return mx.array(
+            [self._bos_token_id] + self._tokenizer.encode("▁" + s.replace(" ", "▁")).ids
+        )
+
+    @property
+    def eos_token_id(self):
+        return self._eos_token_id
+
+    def decode(self, toks: List[int]) -> str:
+        dtoks = self._tokenizer.decode(toks)
+        # TODO, why is this so messed up?
+        return dtoks.replace(" ", "").replace("▁▁", " ").replace("▁", "")
+
+
+def translate_weight_names(name):
+    name = name.replace("blk.", "model.layers.")
+    name = name.replace("ffn_gate", "mlp.gate_proj")
+    name = name.replace("ffn_down", "mlp.down_proj")
+    name = name.replace("ffn_up", "mlp.up_proj")
+    name = name.replace("attn_q", "self_attn.q_proj")
+    name = name.replace("attn_k", "self_attn.k_proj")
+    name = name.replace("attn_v", "self_attn.v_proj")
+    name = name.replace("attn_output", "self_attn.o_proj")
+    name = name.replace("attn_norm", "input_layernorm")
+    name = name.replace("ffn_norm", "post_attention_layernorm")
+    name = name.replace("token_embd", "model.embed_tokens")
+    name = name.replace("output_norm", "model.norm")
+    name = name.replace("output", "lm_head")
+    return name
+
+
+def load(gguf_file: str, repo: str = None):
+    # If the gguf_file exists, try to load model from it.
+    # Otherwise try to download and cache from the HF repo
+    if not Path(gguf_file).exists():
+        if repo is None:
+            raise ValueError(
+                f"Could not find file {gguf_file}, and no Hugging Face"
+                " repo provided for download."
+            )
+        model_path = snapshot_download(
+            repo_id=repo,
+            allow_patterns=[gguf_file],
+        )
+        if not (Path(model_path) / gguf_file).exists():
+            raise ValueError(f"File {gguf_file} not in repo {repo}.")
+        gguf_file = str(Path(model_path) / gguf_file)
+
+    print(f"[INFO] Loading model from {gguf_file}")
+    weights, metadata = mx.load(gguf_file, return_metadata=True)
+    gguf_ft = metadata["general.file_type"]
```

Comment:
Really happy that this now works!

One nit: I have seen GGUF files with mixed quantization formats. I think that the most future proof solution is to export the number of bits and/or group_size as an additional scalar mx.array for each quantized tensor.

Reply:
Do you have any examples of that which are commonly used? If most of the mixing is with q4 and fp16 or q8 and fp16 or quants formats which will be cast to fp16, let's keep it as is for now since it's simple and revisit if it becomes an issue.

It should throw when loading weights if the files have a mixture of q4_0 and q8_0 since the shapes will be wrong, so we can see if anyone files an issue about it.

Reply:
One example is mistral-7b-v0.1.Q4_K_M.gguf from [TheBloke/Mistral-7B-v0.1-GGUF](https://huggingface.co/TheBloke/Mistral-7B-v0.1-GGUF):

```
./gguf-tools show mistral-7b-v0.1.Q4_K_M.gguf | grep blk.0
q4_k tensor blk.0.attn_q.weight @74462848, 16777216 weights, dims [4096,4096], 9437184 bytes
q4_k tensor blk.0.attn_k.weight @83900032, 4194304 weights, dims [4096,1024], 2359296 bytes
q6_k tensor blk.0.attn_v.weight @86259328, 4194304 weights, dims [4096,1024], 3440640 bytes
q4_k tensor blk.0.attn_output.weight @89699968, 16777216 weights, dims [4096,4096], 9437184 bytes
q4_k tensor blk.0.ffn_gate.weight @99137152, 58720256 weights, dims [4096,14336], 33030144 bytes
q4_k tensor blk.0.ffn_up.weight @132167296, 58720256 weights, dims [4096,14336], 33030144 bytes
q6_k tensor blk.0.ffn_down.weight @165197440, 58720256 weights, dims [14336,4096], 48168960 bytes
f32 tensor blk.0.attn_norm.weight @213366400, 4096 weights, dims [4096], 16384 bytes
f32 tensor blk.0.ffn_norm.weight @213382784, 4096 weights, dims [4096], 16384 bytes
````

I think that by default, the quantization util from llama cpp never quantizes `ffn_down` and `attn_v` to less than 6 bits when the quantization format is *K*.

I don't think that's the default behavior for Q4_0, Q4_1 or Q8_0.

Reply:
How do we handle Q4_K_M in the code now? Won't it all just auto cast to 16-bit or will some of it be dequantized as q4_0 in C++?

Reply:
Actually, Q4_K_S mixes Q4_K and Q5_K:

I don't think the difference between K_M and K_S is documented anywhere, but this is probably it. I'm guessing M stands for Medium and S for Small.

```
./gguf-tools show mistral-7b-v0.1.Q4_K_S.gguf\?download=true  | grep blk.0
q4_k tensor blk.0.attn_q.weight @74462848, 16777216 weights, dims [4096,4096], 9437184 bytes
q4_k tensor blk.0.attn_k.weight @83900032, 4194304 weights, dims [4096,1024], 2359296 bytes
q5_k tensor blk.0.attn_v.weight @86259328, 4194304 weights, dims [4096,1024], 2883584 bytes
q4_k tensor blk.0.attn_output.weight @89142912, 16777216 weights, dims [4096,4096], 9437184 bytes
q4_k tensor blk.0.ffn_gate.weight @98580096, 58720256 weights, dims [4096,14336], 33030144 bytes
q4_k tensor blk.0.ffn_up.weight @131610240, 58720256 weights, dims [4096,14336], 33030144 bytes
q5_k tensor blk.0.ffn_down.weight @164640384, 58720256 weights, dims [14336,4096], 40370176 bytes
f32 tensor blk.0.attn_norm.weight @205010560, 4096 weights, dims [4096], 16384 bytes
f32 tensor blk.0.ffn_norm.weight @205026944, 4096 weights, dims [4096], 16384 bytes
```

Reply:
One thing we could do is try and detect if a linear layer is not quantized at the model level. We don't do q5_k so those will be fp16. As long as we don't have a mix of different quantizations (4 bit and 8 bit) that should be pretty easy to do without any downstream changes.

Reply:
> How do we handle Q4_K_M in the code now? Won't it all just auto cast to 16-bit or will some of it be dequantized as q4_0 in C++?

`Q4_K_M` is not actually a tensor type, it's a file type, which mixes `Q4_K` and `Q6_K`. Similarly, `Q4_K_S`, mixes `Q4_K` and `Q5_K `.

Right now we `*K` tensor types to float16, so we don't have this issue of mixed number of bits or block sizes. However, if we add support for Q4_K and Q6_K, it would become a problem.

> One thing we could do is try and detect if a linear layer is not quantized at the model level. 

SGTM.  I don't even need to change anything now, but perhaps we could add a new `.tensor_type` scalar whenever we avoid dequantization, so that we don't need to rely on the `file_type` metadata.